File: OtherStaff/YandexBackendSchool2024/FirstContest/B.cpp
/*
Description:

Input:

Output:

Solution:

*/

#include <iostream>
#include <string>
#include <cctype>
#include <cassert>

using namespace std;

class KeyTrainer {
public:
    void typeChar(char c) {
        typed.insert(cursor, 1, c);
        cursor++;
    }

    void left() {
        if (cursor > 0) {
            cursor--;
        }
    }

    void right() {
        if (cursor < typed.length()) {
            cursor++;
        }
    }

    void bspace() {
        if (cursor > 0) {
            typed.erase(cursor - 1, 1);
            cursor--;
        }
    }

    void delChar() {
        if (cursor < typed.length()) {
            typed.erase(cursor, 1);
        }
    }

    string getTyped() const {
        return typed;
    }

private:
    string typed;
    int cursor = 0;
};


bool isTyped(string a, string log) {
    KeyTrainer kt;
    int i = 0;
    while (i < log.length()) {
        if (isalpha(log[i])) {
            kt.typeChar(log[i]);
            i++;
        } else {
            int j = log.find('>', i);
            string command = log.substr(i + 1, j - i - 1);
            if (command == "bspace") {
                kt.bspace();
            } else if (command == "left") {
                kt.left();
            } else if (command == "right") {
                kt.right();
            } else if (command == "delete") {
                kt.delChar();
            }

            i = j + 1;
        }
    }
    // cout << "str " << kt.getTyped() << "\n";
    return kt.getTyped() == a;
}


void unitTest() {
    {
        string line = "abc";
        string log = "abd<bspace><left><delete>bcd<left><left><delete><right><bspace>c";
        assert(isTyped(line, log));
    }

    {
        string line = "programming";
        string log = "programming<left><left><right><delete>";
        assert(!isTyped(line, log));
    }

    {
        string line = "hellochild";
        string log = "helto<left><bspace>l<delete>ochilds<bspace>";
        assert(isTyped(line, log));
    }
}


int main() {
    unitTest();
    string a, b;
    cin >> a >> b;
    //cout << a << " " << b;
    if (isTyped(a, b)) {
        cout << "Yes\n";
    } else {
        cout << "No\n";
    }
    return 0;
}

File: OtherStaff/YandexBackendSchool2024/FirstContest/B.py
"""
Description:
Keyboard trainer

a - string to type ~ 1000. latin + control keys (delete, bspace, left, right)
b - log of sequence of printed keys or control key

By log check if a string is typed

INPUT:
a
b

OUTPUT
Yes or No

Solution:
1. list of symbols + cursor position
    - cursor points after char that will be deleted by bspace 
    - and points on char that will be deleted by delete
2. process control keys
"""
class KeyTrainer:
    def __init__(self):
        self.log = []
        self.cursor = 0
    
    def typeChar(self, ch):
        self.log.insert(self.cursor, ch)
        self.cursor += 1
    
    def left(self):
        self.cursor -= 1
        if self.cursor < 0:
            self.cursor = 0
    
    def right(self):
        self.cursor += 1
        if self.cursor > len(self.log):
            self.cursor = len(self.log)
    
    def bspace(self):
        if self.cursor > 0:
            self.log.pop(self.cursor - 1)
            self.cursor -= 1
    
    def delete(self):
        if self.cursor < len(self.log):
            self.log.pop(self.cursor)
    
    def getTypedStr(self):
        return "".join(self.log)


def isTyped(toType: str, logged: str):
    kt = KeyTrainer()
    i = 0
    while i < len(logged):
        if logged[i].isalpha():
            kt.typeChar(logged[i])
            i += 1
        else:
            # log[i] == '<'
            comEnd = logged.find('>', i)
            command = logged[i + 1:comEnd]
            if command == "bspace":
                kt.bspace()
            elif command == "delete":
                kt.delete()
            elif command == "left":
                kt.left()
            elif command == "right":
                kt.right()
            i = comEnd + 1

    return kt.getTypedStr() == toType


def main():
    toType = input()
    logged = input()
    if isTyped(toType, logged):
        print("Yes")
    else:
        print("No")

if __name__ == "__main__":
    main()